compiler/src/iree/compiler/Dialect/Flow/IR/FlowOps.cpp

```diff
@@ -364,6 +364,91 @@ ValueRange DispatchRegionOp::getResultDynamicDims(unsigned idx) {
                                shapedType ? shapedType.getNumDynamicDims() : 0);
 }
 
+/// Canonicalizes a DispatchRegionOp: Drop all unused results.
+bool dropUnusedDispatchRegionResults(RewriterBase &rewriter,
+                                     Flow::DispatchRegionOp &regionOp) {
+  assert(regionOp.getBody().hasOneBlock() &&
+         "only regions with one block are supported");
+  OpBuilder::InsertionGuard guard(rewriter);
+  rewriter.setInsertionPoint(regionOp);
+
+  // Determine unused results and result types + dynamic dimensions of the new
+  // op.
+  llvm::DenseSet<unsigned> unusedResults;
+  SmallVector<Type> resultTypes;
+  SmallVector<Value> dynamicDims;
+  unsigned dimOffset = 0;
+  for (const auto &it : llvm::enumerate(regionOp.getResults())) {
+    Type type = it.value().getType();
+    auto shapedType = type.dyn_cast<ShapedType>();
+    if (it.value().use_empty()) {
+      unusedResults.insert(it.index());
+    } else {
+      resultTypes.push_back(type);
+      ValueRange dims = regionOp.getResultDims().slice(
+          dimOffset, shapedType.getNumDynamicDims());
+      dynamicDims.append(dims.begin(), dims.end());
+    }
+    dimOffset += shapedType.getNumDynamicDims();
+  }
+  assert(dimOffset == regionOp.getResultDims().size() &&
+         "expected that all dynamic dims were processed");
+
+  // Nothing to do if all results are used.
+  if (unusedResults.empty()) return false;
+
+  // Create new region and move over the body.
+  auto newRegionOp = rewriter.create<Flow::DispatchRegionOp>(
+      regionOp.getLoc(), resultTypes, dynamicDims);
+  newRegionOp.getBody().takeBody(regionOp.getBody());
+
+  // Update terminator.
+  auto returnOp =
+      cast<Flow::ReturnOp>(newRegionOp.getBody().front().getTerminator());
+  SmallVector<Value> yieldedValues;
+  for (const auto &it : llvm::enumerate(returnOp.getOperands()))
+    if (!unusedResults.contains(it.index()))
+      yieldedValues.push_back(it.value());
+  rewriter.updateRootInPlace(
+      returnOp, [&]() { returnOp.operandsMutable().assign(yieldedValues); });
+
+  // Replace all uses of the old op.
+  unsigned resultCounter = 0;
```

Comment:
Easier to just create a `SmallVector<Value> replacements` of the same length as the number of results of the original operation initialized to `nullptr`. For things that are unused in the original op and are dropped, leave it nullptr. Replace entries corresponding to result in the old op with the corresponding result from the new op. Then you can just do `replaceOp(regionOp, replacements)`